jax/_src/util.py

```diff
@@ -300,16 +300,29 @@ def wrapper(*args, **kwargs):
 
     wrapper.cache_clear = cached.cache_clear
     wrapper.cache_info = cached.cache_info
-    cache_clearing_funs.add(wrapper.cache_clear)
+    register_cache(wrapper, str(f))
     return wrapper
   return wrap
 
-cache_clearing_funs = weakref.WeakSet()  # type: ignore
+# Maps caches to the name of the callable they apply to. All caches in
+# this dictionary support `cache_clear()`.
+_caches: dict[Any, str] = {}
```

Comment:
Shouldn't this be a WeakSet? Why make it a dict?

Reply:
This should be indeed a `weakref.WeakKeyDictionary`, but the reason I cannot make it that just yet, is that there are some references to the function `backend.add_clear_backends_callback` in Google internal code. I was planning to make those changes in a new PR, but on second thought I will make them as part of landing this.

Reply:
The name was very important for my debugging. I am not done debugging cache leaks, there is quite a bit more to do, and I will need it again. I just wanted to land these cleanups and check for failures before landing bigger caching changes.

Reply:
Ok, sounds good. But please make it a WeakDictionary